# Assignment Week 7

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  #library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})

ah <- AnnotationHub(localHub=TRUE)
ensdb <- ah[["AH89211"]] # mouse ensembldb object
```

## Preparing the tracks
```{r}
options(timeout = 1800)
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam", mode = "wb")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai", mode = "wb")
```
```{r}
bam <- "atac.chr19.bam"

# big wig has only summary information for different positions of the genome, bam is a bigger file
# create a track using all fragments, computing the coverage at each position (and force Ensembl naming)
bam2bw(bam, output_bw = "full_cov.bw", paired=TRUE, binWidth=10L, forceSeqlevelsStyle = "Ensembl")

# create a track using all fragments, but computing the number of cuts/insertion sites at each position
epiwraps::bam2bw(bam, output_bw = "full_cuts.bw", paired=TRUE, binWidth=1L, type="ends", #counts only ends
                 shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")

# create a track using only nucleosome-free fragments, computing coverage
bam2bw(bam, output_bw = "NF_cov.bw", paired=TRUE, binWidth=5L, minFragLength=30,
       maxFragLength=120, forceSeqlevelsStyle = "Ensembl")

# create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=30,
       maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")

# create a track using only mono-nucleosome fragments, computing coverage
bam2bw(bam, output_bw = "mono_cov.bw", paired=TRUE, binWidth=10L, minFragLength=140,
       maxFragLength=220, forceSeqlevelsStyle = "Ensembl")

# create a track using only mono-nucleosome fragments, computing centers
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140,
       maxFragLength=220, type="center", forceSeqlevelsStyle = "Ensembl")
```

## Factor KLF4
### Get KLF4 motif
```{r}
motif_KLF4 <- query(MotifDb, c("KLF4","Mus"))[[1]]
motif2_KLF4 <- convert_motifs(motif_KLF4, class = "TFBSTools-PFMatrix")
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif occurences
moi_KLF4 <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out = "positions", p.cutoff = 1e-5)[[1]]
moi_KLF4 <- as(setNames(moi_KLF4, names(chr19)), "GRanges")
```

### Plotting signal around KLF4 motif occurences
```{r}
tracks <- c("Full coverage"="full_cov.bw", "NF coverage"="NF_cov.bw", "All cuts"="full_cuts.bw",
            "NF cuts"="NF_cuts.bw", "Mono cov"="mono_cov.bw", "Mono centers"="mono_centers.bw")
sm_KLF4 <- signal2Matrix(tracks, moi_KLF4, w = 5, extend = 500)
nf <- bwNormFactors(tracks, useSeqLevels="19", nwind=5000L)
smb_KLF4 <- rescaleSignalMatrices(sm_KLF4, nf)
plotEnrichedHeatmaps(smb_KLF4, trim = 0.95, minRowVal = 15, colors = c("white", "darkred"))
```

## Factor MAZ
### Get MAZ motif
```{r}
motif_MAZ <- query(MotifDb, c("MAZ","Mus"))[[1]]
motif2_MAZ <- convert_motifs(motif_MAZ, class = "TFBSTools-PFMatrix")

# find motif occurences
moi_MAZ <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out = "positions", p.cutoff = 1e-5)[[1]]
moi_MAZ <- as(setNames(moi_MAZ, names(chr19)), "GRanges")
```

### Plotting signal around MAZ motif occurences
```{r}
sm_MAZ <- signal2Matrix(tracks, moi_MAZ, w = 5, extend = 500)
smb_MAZ <- rescaleSignalMatrices(sm_MAZ, nf)
plotEnrichedHeatmaps(smb_MAZ, trim = 0.95, minRowVal = 15, colors = c("white", "darkred"))
```