---
title: "R Notebook"
output: html_notebook
---
# Assignment Week 6
```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  #library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives
  library(motifmatchr)
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs
  library(TFBSTools) # for handling some motif formats
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis
})
# loading function findMotifInstances
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the
#' coordinate conversion as `memes` does.
#'
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with
#'   coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i],
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```

## Downlaoding peaks for transcription factor EGR1 in humans
```{r}
dir.create("egr1")
download.file("https://www.encodeproject.org/files/ENCFF375QIT/@@download/ENCFF375QIT.bed.gz", "egr1/EGR1.bed.gz")
peaksEGR1 <- rtracklayer::import("egr1/EGR1.bed.gz", format = "NarrowPeak")
seqlevelsStyle(peaksEGR1) <- "Ensembl"
peaksEGR1_chr1 <- peaksEGR1[seqnames(peaksEGR1)=="1"]
```

## Identifying the instances of the factor's motif
```{r}
peaksEGR1_centers <- resize(peaksEGR1_chr1, fix="center", width=100)
ah <- AnnotationHub()
genome <- ah[["AH5012"]]
peaksEGR1_seqs <- Biostrings::getSeq(genome, peaksEGR1_centers_centers)
names(peaksEGR1_seqs) <- as.character(granges(peaksEGR1_centers))
```

```{r}
motifsEGR1 <- query(MotifDb, "EGR1")
names(motifsEGR1)
motifEGR1 <- motifsEGR1[["Hsapiens-jaspar2022-EGR1-MA0162.4"]]
view_motifs(motifEGR1)

```

## Answers to questions
### Of all the peaks what proportion contains a motif for the factor?
```{r}
Biostrings::writeXStringSet(rtracklayer::import(genome), "egr1/genome.fa")
#motifEGR1_2 <- convert_motifs(motifEGR1, class = "TFBSTools-PFMatrix")
#motifEGR1_3 <- matchMotifs(motifEGR1_2, subject = peaksEGR1_centers, genome = Rsamtools::FaFile("genome.fa"), out = "positions")
motifEGR1_2 <- findMotifInstances(peaksEGR1_seqs, motifEGR1)
peaks_w_motif <- peaksEGR1_chr1[overlapsAny(peaksEGR1_chr1, motifEGR1_2)]
length(peaksEGR1_chr1)
length(peaks_w_motif)/length(peaksEGR1_chr1)
```
Of the XX peaks, XX (XX%) contain a motif.

### Of all instances of that motif in the genome (or in one chromosome), what proportion is bound by the factor (i.e. has a peak)?
```{r}
motifsEGR1_3 <- findMotifInstances(genome, motifEGR1)
motif_w_peaks <- genome[overlapsAny(genome, motifsEGR1_3)]
length(genome)
length(motif_w_peaks)/length(genome)
```
Of the XX motif instances, XX (XX%) overlap a peak.